### PR TITLE
feat: cloudformation condition chaining

### LINKED ISSUE #1457: Cannot use `Condition` (or `condition.ref`) in `FnOr`
eg:

```yaml
    HasManagedPolicies:
      !Or [ { Condition: UseManagedPolicyARN }, { Condition: UseECR } ]
```

or even:

```yaml
    CreateMetricsStack:
      Condition: UseAutoscaling
```

It *feels* like I should be able to do this:

```typescript

    const useManagedPolicyARN = new cdk.Condition(...);
    const useEcr = new cdk.Condition(...);
    const hasManagedPolicies = new cdk.Condition(this, "HasManagedPolicies", {
      expression: new cdk.FnOr(
        useManagedPolicyARN,
        useEcr
      )
    });
```

But the `cdk.Condition` class does not match the type definition of the predicate functions (which require `FnCondition`.

After some mucking around, I found the following workaround, but it relies on `Fn` simply re-emitting exactly what it's passed:

```typescript
    const hasManagedPolicies = new cdk.Condition(this, "HasManagedPolicies", {
      expression: new cdk.FnOr(
        new cdk.Fn("Condition", "UseManagedPolicyARN"),
        new cdk.Fn("Condition", "UseECR")
      )
    });
```

This applies to referring to Conditions within `FnOr` and so forth within resources as well.

What is the correct way to express this in CDK?

### PR DESCRIPTION
**Pull request by eladb**
Change `conditionXx` methods to accept an interface `IConditionExpression`
instead of concrete class and implement this interface by the `Condition`
construct.

This enables chaining conditions like so:

    const cond1, cond2, cond, cond4 = new cdk.Condition(...)
    Fn.conditionOr(cond1, cond2, Fn.conditionAnd(cond3, cond4))

Fixes #1457

----

### Pull Request Checklist

* [x] Testing
  - Unit test added
  - __CLI change?:__ manually run [integration tests](https://github.com/awslabs/aws-cdk/blob/master/packages/aws-cdk/integ-tests/test.sh) and paste output as a PR comment
  - __cdk-init template change?:__ coordinated update of integration tests with team
* [x] Docs
  - __jsdocs__: All public APIs documented
  - __README__: README and/or documentation topic updated
* [x] Title and Description
  - __Change type__: title prefixed with **fix**, **feat** will appear in changelog
  - __Title__: use lower-case and doesn't end with a period
  - __Breaking?__: last paragraph: "BREAKING CHANGE: <describe what changed + link for details>"
  - __Issues__: Indicate issues fixed via: "**Fixes #xxx**" or "**Closes #xxx**"
* [x] Sensitive Modules (requires 2 PR approvers)
  - IAM Policy Document (in @aws-cdk/aws-iam)
  - EC2 Security Groups and ACLs (in @aws-cdk/aws-ec2)
  - Grant APIs (only if not based on official documentation with a reference)

----

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license.

### CODE REVIEW

File: packages/@aws-cdk/cdk/lib/cloudformation/condition.ts [hunk 1]
```diff
@@ -26,35 +25,52 @@ export class Condition extends Referenceable {
...
-export class FnCondition extends CloudFormationToken {
-  constructor(type: string, value: any) {
-    super({ [type]: value });
-  }
-}
+export interface IConditionExpression {
+  /**
+   * Returns a JSON node that represents this condition expression
+   */
+  resolve(): any;
```
Comment (rix0rrr) [inline packages/@aws-cdk/cdk/lib/cloudformation/condition.ts]: Why `resolve()` and not a more specific fucntion name? If you wanted `resolve()` you could have accepted a Token.

File: packages/@aws-cdk/cdk/lib/cloudformation/fn.ts [hunk 2]
```diff
@@ -442,13 +442,23 @@ class FnCidr extends FnBase {
   }
 }
 
+class FnConditionBase extends CloudFormationToken implements IConditionExpression {
+  constructor(type: string, value: any) {
+    super({ [type]: value });
+  }
+
+  public toConditionJson() {
```
Comment (rix0rrr) [inline packages/@aws-cdk/cdk/lib/cloudformation/fn.ts]: Where is this called?

File: packages/@aws-cdk/cdk/lib/cloudformation/condition.ts (hunk 1 above)
Comment (eladb) [reply packages/@aws-cdk/cdk/lib/cloudformation/condition.ts]: But I wanted the Condition construct to implement it, so I had to define it as an interface. And I want the recursive resolution behavior of tokens to apply to full expressions, so I just faked it.
We should probably define an `IToken` with a resolve method so (reopened #79)

File: packages/@aws-cdk/cdk/lib/cloudformation/fn.ts (hunk 2 above)
Comment (eladb) [reply packages/@aws-cdk/cdk/lib/cloudformation/fn.ts]: oops. will remove

**Pull request merged**